backend/server/server.go

```diff
@@ -68,343 +83,58 @@ func HomeHandler(w http.ResponseWriter, req *http.Request) {
-		util.LogCall(req.Method, req.RequestURI, "200")
-		return
-	}
-
-	// Serve from cache instead
-	fileLocation := fmt.Sprintf("../data/%s/pullRequests.json", vars["username"])
-	content, err := ioutil.ReadFile(fileLocation)
-	if err != nil {
-		w.Header().Set("Content-Type", "application/json")
-		w.WriteHeader(http.StatusUnauthorized)
-		res := response{
-			Status: "401",
-			Body:   fmt.Sprint(err),
-		}
-		json.NewEncoder(w).Encode(res)
-
-		util.LogCall(req.Method, req.RequestURI, "401")
-		return
-	}
-
-	w.Header().Set("Content-Type", "application/json")
-	fmt.Fprintf(w, string(content))
-	util.LogCall(req.Method, req.RequestURI, "200")
-
+	util.LogCall(req.Method, req.RequestURI, "200", vars["startTime"], false)
 }
 
-// GetFellowIssuesCreated get a list of the recent issues created by a user
-func GetFellowIssuesCreated(w http.ResponseWriter, req *http.Request) {
+func Query(w http.ResponseWriter, req *http.Request) {
 	vars := mux.Vars(req)
 
-	if !util.CheckUser(vars["username"], "issuesCreated.json") {
-
-		httpClient := util.SetupOAuth()
-		client := graphql.NewClient("https://api.github.com/graphql", httpClient)
+	// Call the GitHub API and cache the result
+	if !util.CacheExists(vars["username"], vars["fileName"]) {
 
-		var tempStruct queries.MegaJSONStruct
-
-		variables := map[string]interface{}{
-			"username": graphql.String(vars["username"]),
+		client := util.SetupOAuth()
+		dataStruct, variables := util.GetStruct(vars["query"], vars["username"])
+		if dataStruct == nil {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusUnauthorized)
+			res := response{
+				Status: "401",
+				Body:   fmt.Sprint("Invalid query type given"),
+			}
+			json.NewEncoder(w).Encode(res)
+			util.LogCall(req.Method, req.RequestURI, "401", vars["startTime"], false)
+			return
 		}
 
-		// Call the API
-		err := client.Query(context.Background(), &tempStruct.IssCreated, variables)
-		util.CheckAPICallErr(err)
-
-		// Write to JSON file
-		dirLocation := fmt.Sprintf("../data/%s", vars["username"])
-		_ = os.Mkdir(dirLocation, 0755)
-
-		fileLocation := fmt.Sprintf("../data/%s/issuesCreated.json", vars["username"])
-		jsonData, err := json.Marshal(tempStruct.IssCreated)
+		err := client.Query(context.Background(), dataStruct, variables)
+		err = util.CheckAPICallErr(err)
 		if err != nil {
-			log.Fatal(err)
+			// This catches errors thrown due to invalid usernames which is rare if not caught by the
+			// verification middleware
+			match, err := regexp.MatchString(`(Could not resolve to a User  with the login of \')(.)+(\')`, err.Error())
+			if err != nil && !match {
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(http.StatusUnauthorized)
+				res := response{
+					Status: "401",
+					Body:   fmt.Sprint("Invalid username given"),
+				}
+				json.NewEncoder(w).Encode(res)
+				util.LogCall(req.Method, req.RequestURI, "401", vars["startTime"], false)
+				return
```

Comment:
Same as above, this could be a function `UnauthorizedResponse` that takes the error message (in this case "Invalid username given", in the previous case "Invalid query type given").